### DISCUSSION TITLE
Monai Label Training

### DISCUSSION THREAD

**Discussion by stevenagl12** [Project-MONAI/MONAILabel#1055] [Q&A]
I have a folder for imagesTr with my training images, and a subfolder for labels with nrrd files for my labelmaps. How can I train a model with this using monai label?

**Comment by diazandr3s** [comment]
Hi @stevenagl12,

This is a good question. 

If working on a local PC, MONAI Label studies should follow this structure: https://youtu.be/wtiEe_jiUzg?t=3267

You may find this video useful as well: https://www.youtube.com/watch?v=3HTh2dqZqew&list=PLtoSVSQ2XzyD4lc-lAacFBzOdv5Ou-9IA&index=3

Hope this helps

**Reply by stevenagl12** [reply]
I apologize for the late reply, but neither of these really help to solve the question. So I have my files in the correct format already from the first video. How do I train a new model now given that I have this labeled data? All the videos I can find say how to train amodel from scratch with unlabeled data using the segment editor process of the scribbles processes. Or viceversa how to download a trained model and infer with the testing data, but how do I run training with the data that is in the desired format?

Another way to ask this is how do we train a new model after editing the deepedit config file with the appropriate class dictionary and turning off the pretrained model option (while removing the files for any models from the model folder) from the Task01 or Task04 training data?

**Reply by diazandr3s** [reply]
Thanks for the update, @stevenagl12.

> Another way to ask this is how do we train a new model after editing the deepedit config file with the appropriate class dictionary and turning off the pretrained model option (while removing the files for any models from the model folder) from the Task01 or Task04 training data?

You've already done the most difficult part. To trigger the training process, just click on the Train button in the 3D Slicer module shown here:
![train-](https://user-images.githubusercontent.com/11991079/198155652-63a916cc-fe9b-4329-b464-dd431369006f.png)

You can find more information here: https://docs.monai.io/projects/label/en/latest/quickstart.html

Hope that helps,

**Comment by stevenagl12** [comment]
How would you get the training and testing curves from the process then?

**Reply by diazandr3s** [reply]
MONAI Label uses tensorboard to keep track of training and validation curves. You can find those tensorboard files in the model folder once training starts.

**Reply by stevenagl12** [reply]
Ok, so another question is for the Task01 Brain Tumour segmentation, is there preprocessing that should take place in order to get the system to work? I am getting a size mismatch error that says: "RuntimeError: Given groups=1, weight of size [32, 5, 3, 3, 3], expected input[1, 8, 128, 128, 128] to have 5 channels, but got 8 channels instead."

**Reply by SachidanandAlle** [reply]
You might be using pretrained model that was done over 8 channels.. but your input has only 5 of them..

Fix the number of channels based on your data and discard existing pretrained model and train from scratch.. or if you have only subset of 8.. add empty channels to for the missing labels and fine tune model for your data

**Comment by stevenagl12** [comment]
I'm very confused with what I am doing wrong here. When I first connect to
Monailabel for the server, the Train button is blacked out. As I said
before, I have edited the config file, and removed all of the models. How
am I supposed to train any new models from existing data when the train
button isn't even active?

On Thu, Oct 27, 2022 at 12:59 PM SACHIDANAND ALLE ***@***.***>
wrote:

> You might be using pretrained model that was done over 8 channels.. but
> your input has only 5 of them..
>
> Fix the number of channels based on your data and discard existing
> pretrained model and train from scratch.. or if you have only subset of 8..
> add empty channels to for the missing labels and fine tune model for your
> data
>
> —
> Reply to this email directly, view it on GitHub
> <https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FProject-MONAI%2FMONAILabel%2Fdiscussions%2F1055%23discussioncomment-3984173&data=05%7C01%7Csalewis%40g-mail.buffalo.edu%7C84cce36b7e764a225f3d08dab83ca5e4%7C96464a8af8ed40b199e25f6b50a20250%7C0%7C0%7C638024867855200980%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=UuV2cNiemh1dLhcdO%2Bu3d5Knu70klVLGoIFcQPWz9rI%3D&reserved=0>,
> or unsubscribe
> <https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAHHKCLLP33R3F4GLKMHDOBDWFKYH7ANCNFSM6AAAAAARAAE4FM&data=05%7C01%7Csalewis%40g-mail.buffalo.edu%7C84cce36b7e764a225f3d08dab83ca5e4%7C96464a8af8ed40b199e25f6b50a20250%7C0%7C0%7C638024867855200980%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=in%2Ft0TtJpTlBdQpbNb9bcbL2e6%2FXXbIQnYN5%2BfR5%2Bkw%3D&reserved=0>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Sincerely,

Steven A. Lewis

**Reply by SachidanandAlle** [reply]
2 situations..
1. You don't have any labeled data for training 
2. You don't have training task or disabled it

You can check server:port/info.. the training tasks and how much data is labeled is shown there..

**Reply by stevenagl12** [reply]
So, I'm wondering if I am missing something critical here. I am using the tasked datasets so I know that the training data has labels for them. I am running the following command monailabel start_server -a apps/radiology -s datasets/Task04_Hippocampus/imagesTr -c models deepedit. I can access the individual nifti files through 3D Slicer, but the train option is blacked out unless I annotate a specific sample and then submit it, then the train option is able to be pressed. This isn't what I want as this would train from a single sample and I am also not familiar enough with mri imaging to be able to correctly annotate every sample. I want to create a new trained model from the Task01 or Task04 datasets that have the labeled data, then run inference to see how the trained model performs. How do I go about doing that?

As I mentioned I know to edit the config file for deepedit for the necessary classes, to turn off the option for using a pretrained model:         if strtobool(self.conf.get("use_pretrained_model", "false")):
            url = f"{self.conf.get('pretrained_path', self.PRE_TRAINED_PATH)}/deepedit_{network}_multilabel.pt"
            download_file(url, self.path[0]),
            
and I know that if I want to work with the BRATS dataset I need to change the in_channels to be 4, but other than that how to I get the model to train?

**Reply by SachidanandAlle** [reply]
You just need to moved the existing labels into {studies}/labels/final

**Reply by stevenagl12** [reply]
Ok, this wasn't in any of the videos I had reviewed so hopefully this is the problem.

**Reply by SachidanandAlle** [reply]
May be later you can help so improve the documentation part.. may be a wiki etc is also fine.. or some section on how to...?

**Reply by stevenagl12** [reply]
If this works, I will absolutely do that.

**Reply by diazandr3s** [reply]
> Ok, this wasn't in any of the videos I had reviewed so hopefully this is the problem.

Hi @stevenagl12,

This folder structure was exactly the one I suggested in the first comment: https://github.com/Project-MONAI/MONAILabel/discussions/1055#discussioncomment-3830237

Here is again for your reference: https://youtu.be/wtiEe_jiUzg?t=3275

Hope this helps,

**Reply by stevenagl12** [reply]
The problem wasn't the folder structure. It was knowing that the data structure that is downloaded from the published tasks has the data in the incorrect folder to be able to train. The necessary labels for the data are in the labelsTr folder outside of the imagesTr or imagesTs folders, while the labels folder with the final and original folders in the imagesTr folder were empty for both of these tasks. I thought the dataset.json file that was downloaded that contained all of the necessary paths for the labels was sufficient to be able to run the training. Still having training errors but less so and working on those now.
